src/buffer/out/Row.hpp

```diff
@@ -20,74 +20,139 @@ Revision History:
+    void TransferAttributes(const til::small_rle<TextAttribute, uint16_t, 1>& attr, til::CoordType newWidth);
+
+    void ClearCell(til::CoordType column);
+    OutputCellIterator WriteCells(OutputCellIterator it, til::CoordType columnBegin, std::optional<bool> wrap = std::nullopt, std::optional<til::CoordType> limitRight = std::nullopt);
+    bool SetAttrToEnd(til::CoordType columnBegin, TextAttribute attr);
+    void ReplaceCharacters(til::CoordType columnBegin, til::CoordType width, const std::wstring_view& chars);
+
+    const til::small_rle<TextAttribute, uint16_t, 1>& Attributes() const noexcept;
+    TextAttribute GetAttrByColumn(til::CoordType column) const;
+    std::vector<uint16_t> GetHyperlinks() const;
+    uint16_t size() const noexcept;
+    til::CoordType MeasureLeft() const noexcept;
+    til::CoordType MeasureRight() const noexcept;
+    bool ContainsText() const noexcept;
+    std::wstring_view GlyphAt(til::CoordType column) const noexcept;
+    DbcsAttribute DbcsAttrAt(til::CoordType column) const noexcept;
+    std::wstring_view GetText() const noexcept;
+    DelimiterClass DelimiterClassAt(til::CoordType column, const std::wstring_view& wordDelimiters) const noexcept;
+
+    auto AttrBegin() const noexcept { return _attr.begin(); }
+    auto AttrEnd() const noexcept { return _attr.end(); }
 
 #ifdef UNIT_TESTING
     friend constexpr bool operator==(const ROW& a, const ROW& b) noexcept;
     friend class RowTests;
 #endif
 
 private:
-    CharRow _charRow;
-    ATTR_ROW _attrRow;
-    LineRendition _lineRendition;
-    til::CoordType _id;
-    til::CoordType _rowWidth;
+    static constexpr uint16_t CharOffsetsTrailer = 0x8000;
+    static constexpr uint16_t CharOffsetsMask = 0x7fff;
+
+    template<typename T>
+    static constexpr uint16_t _clampedUint16(T v) noexcept;
+    template<typename T>
+    constexpr uint16_t _clampedColumn(T v) const noexcept;
+    template<typename T>
+    constexpr uint16_t _clampedColumnInclusive(T v) const noexcept;
+
+    wchar_t _uncheckedChar(size_t off) const noexcept;
+    uint16_t _charSize() const noexcept;
+    uint16_t _uncheckedCharOffset(size_t col) const noexcept;
+    bool _uncheckedIsTrailer(size_t col) const noexcept;
+
+    void _init() noexcept;
+    void _resizeChars(uint16_t colExtEnd, uint16_t chExtBeg, uint16_t chExtEnd, size_t chExtEndNew);
+
+    // These fields are a bit "wasteful", but it makes all this a bit more robust against
+    // programming errors during initial development (which is when this comment was written).
+    // * _chars and _charsHeap are redundant
+    //   If _charsHeap is stored in _chars, we can still infer that
+    //   _chars was allocated on the heap if _chars != _charsBuffer.
+    // * _chars doesn't need a size_t size()
+    //   The size may never exceed an uint16_t anyways.
+    // * _charOffsets doesn't need a size() at all
+    //   The length is already stored in _columns.

```

Comment:
I personally wouldn't say so. IMO it's not something we need to try and pursue any time soon, because there's many much better opportunities to reduce our memory usage. I only left the comment so that a future reader will have an easier time optimizing this code if we ever need to.